File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(string s) {
        int answer = 0;
                
        for(int i = 0; i < s.Length; i++)
        {
            if(CheckCorrect(s))
            {
                answer++;
            }
            s = Shift(s);
        }
        return answer;
    }
    private bool CheckCorrect(string s)
    {
        Stack<char> stack = new Stack<char>();
        
        for(int i = 0; i < s.Length; i++)
        {
            if(s[i] == '(' || s[i] == '[' || s[i] == '{')
            {
                stack.Push(s[i]);                              
            }
            else
            {
                if(stack.Count == 0) return false;
                
                if(stack.Peek() == '(' && s[i] == ')')
                {
                    stack.Pop();
                }
                else if(stack.Peek() == '[' && s[i] == ']')
                {
                    stack.Pop();
                }
                else if(stack.Peek() == '{' && s[i] == '}')
                {
                    stack.Pop();
                }
                else return false;
            }
        }
        return stack.Count == 0;
    }
               
    
    private string Shift(string s)
    {
        return s.Substring(1) + s[0];
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

bool check(string s)
{
    stack<char> stk;
    
    for(char c : s)
    {
        if(c == '(' || c == '[' || c == '{')
        {
            stk.push(c);
        }
        else
        {
            if(!stk.size()) return false;
            
            if(c == ')' && stk.top() != '(') return false;
            else if(c == ']' && stk.top() != '[') return false;
            else if(c == '}' && stk.top() != '{') return false;
            
            stk.pop();
        }            
    }
    return stk.size() ? false : true;
}
int solution(string s) {
    int answer = 0;
    
    for(int i = 0; i < s.size(); i++)
    {
        string rotated = s.substr(i) + s.substr(0, i);
        if(check(rotated))
            answer++;
    }
    return answer;
}